lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,328 @@
+/*
+Copyright 2020 Gravitational, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package benchmark
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// minValue is the min millisecond recorded for histogram
+	minValue = 1
+	// maxValue is the max millisecond recorded for histogram
+	maxValue = 60000
+	// significantFigures is the precision of the values
+	significantFigures = 3
+	// pauseTimeBetweenBenchmarks is the time to pause between each benchmark
+	pauseTimeBetweenBenchmarks = time.Second * 5
+)
+
+// Config specifies benchmark requests to run
+type Config struct {
+	// Threads is amount of concurrent execution threads to run
+	Threads int
+	// Rate is requests per second origination rate
+	Rate int
+	// Command is a command to run
+	Command []string
+	// Interactive turns on interactive sessions
+	Interactive bool
+	// MinimumWindow is the min duration
+	MinimumWindow time.Duration
+	// MinimumMeasurments is the min amount of requests
+	MinimumMeasurements int
+}
+
+// Result is a result of the benchmark
+type Result struct {
+	// RequestsOriginated is amount of reuqests originated
+	RequestsOriginated int
+	// RequestsFailed is amount of requests failed
+	RequestsFailed int
+	// Histogram is a duration histogram
+	Histogram *hdrhistogram.Histogram
+	// LastError contains last recorded error
+	LastError error
+	// Duration it takes for the whole benchmark to run
+	Duration time.Duration
+}
+
+// Run is used to run the benchmarks, it is given a generator, command to run,
+// a host, host login, and proxy. If host login or proxy is an empty string, it will
+// use the default login
+func Run(ctx context.Context, lg *Linear, cmd, host, login, proxy string) ([]Result, error) {
+	var results []Result
+	sleep := false
```

Comment:
nit: move these down, right before the loop, as close to their usage location as possible
that way a reader won't have to jump up and down the function to see where vars come from